File: src/main/java/com/leetcode/aprilchallenge/NumberOfIslands.java
package com.leetcode.aprilchallenge;

/**
 * @author wezhyn
 * @since 04.18.2020
 */
public class NumberOfIslands {

    public int numIslands(char[][] grid) {
        int num = 0;
        if (grid == null) {
            return num;
        }
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                num += islands(grid, i, j, grid.length - 1, grid[i].length - 1);
            }
        }
        return num;
    }

    private int islands(char[][] grid, int m, int n, int maxLine, int maxColumn) {
        int result = 0;
        if (m < 0 || n < 0 || m > maxLine || n > maxColumn) {
            return 0;
        }
        if (grid[m][n] == '1') {
            result = 1;
            grid[m][n] = 0;
            islands(grid, m + 1, n, maxLine, maxColumn);
            islands(grid, m - 1, n, maxLine, maxColumn);
            islands(grid, m, n + 1, maxLine, maxColumn);
            islands(grid, m, n - 1, maxLine, maxColumn);
        } else {
            return result;
        }
        return result;
    }
}

File: src/main/java/com/leetcode/graph/NumberOfIslands.kt
package com.leetcode.graph

/**
 *
 *
 * Given a 2d grid map of '1's (land) and '0's (water), count the number of islands.
 * An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
 * You may assume all four edges of the grid are all surrounded by water.
 * Example 1:
 * Input:
 * 11110
 * 11010
 * 11000
 * 00000
 *
 * Output: 1
 *
 * 使用 dfs 将岛屿变成 0
 * @author wezhyn
 * @since 06.19.2020
 *
 */
class NumberOfIslands {
    fun numIslands(grid: Array<CharArray>): Int {
        var num = 0
        for ((r, row) in grid.withIndex()) {
            for (l in row.indices) {
                if (row[l] == '1') {
                    num++
                    dfs(grid, r, l)
                }
            }
        }
        return num
    }

    private fun dfs(grid: Array<CharArray>, r: Int, l: Int) {
        if (r >= grid.size || l >= grid[0].size
                || r < 0 || l < 0
                || grid[r][l] == '0') {
            return
        }
        grid[r][l] = '0'
        dfs(grid, r, l - 1)
        dfs(grid, r, l + 1)
        dfs(grid, r + 1, l)
        dfs(grid, r - 1, l)
    }
}